flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java

```diff
@@ -222,19 +225,65 @@ public SingleInputGate(
 	}
 
 	@Override
-	public void setup() throws IOException, InterruptedException {
+	public void setup() throws IOException {
 		checkState(this.bufferPool == null, "Bug in input gate setup logic: Already registered buffer pool.");
 		// assign exclusive buffers to input channels directly and use the rest for floating buffers
 		assignExclusiveSegments();
 
 		BufferPool bufferPool = bufferPoolFactory.get();
 		setBufferPool(bufferPool);
+	}
 
-		requestPartitions();
+	@Override
+	public void initializeStateAndRequestPartitions(
+			boolean hasStates,
+			@Nullable ExecutorService executor,
+			ChannelStateReader reader) throws Exception {
+
+		if (hasStates) {
+			checkNotNull(executor);
+			readRecoveredStateBeforeRequestPartition(executor, reader);
+		} else {
+			requestPartitions();
+		}
```

Comment:
This doesn't look right - boolean `hasStates` and `@Nullable executor` - it looks like this check should be on a different layer. As this method looks like it's called only in one place, shouldn't it be just inlined?

Reply:
It actually inlines two requirements on purpose. TBH I was also a bit torn to do so before.

My previous thought was trying not to affect the normal process if not enabling unaligned mode, that means requesting partitions directly, otherwise short lifecycle jobs might be sensitive for the potential performance regression with unnecessary state recovery process. And passing `nullable` executor can also simplify the unit tests without unaligned mode not to maintain the useless `executor`.

Another solution is to separate it into two methods. One is for `#requestPartitions` and the other is for `readRecoveredState`. But that means we also need to define another explicit interface method `InputGate#requestPartitions`, which would bring more tiny steps in the lifecycle of `InputGate`, like `setup` -> `readRecoveredState` -> `requestPartitions`. So I chose the inlined way to reduce this overhead.

Reply:
Ok, I see. This is messy :/ It comes down to this issue that input gates are initialised in the `Task` and state is handled in `StreamTask` :/ It should have been handled together upon the construction of `SingleInputGate` class.

I think I would be +0.1 for dropping `initializeStateAndRequestPartitions` and both `readRecoveredStateBeforeRequestPartition` and `requestPartitions` to the interfaces, but if you think it's better to have a single method, I'm fine with that.

Reply:
I am planning to separate this inlined method for your 0.1 favor. :)